File: TierSpec/TierSpec/Models/StateMachine.swift
//
//  StateMachine.swift
//  TierSpec
//
//  Created by Sisyphus on 2026/4/26.
//

import Foundation

// MARK: - Actor Type

/// Represents the type of actor performing a state transition
enum ActorType: String, Codable, CaseIterable {
    case human
    case ai
    case system
}

/// Represents the valid transitions in the SDLC workflow
enum StateMachine {
    
    // MARK: - Allowed Transitions
    
    /// Returns the allowed target states for a given current state
    static func allowedTransitions(from currentState: ItemStatus, actorType: ActorType = .human) -> [ItemStatus] {
        if currentState == .cancelled {
            return []
        }
        
        if currentState == .done {
            return [.blocked, .cancelled, .needs_info]
        }
        
        if currentState == .blocked {
            return [.todo, .in_progress, .cancelled, .needs_info]
        }
        
        if currentState == .needs_info {
            return [.todo, .blocked, .cancelled]
        }
        
        let baseTransitions = baseAllowedTransitions(from: currentState, actorType: actorType)
        return baseTransitions + [.blocked, .cancelled, .needs_info]
    }
    
    private static func baseAllowedTransitions(from state: ItemStatus, actorType: ActorType) -> [ItemStatus] {
        switch state {
        case .todo:
            return [.in_progress]
        case .in_progress:
            return [.test]
        case .test:
            let transitions: [ItemStatus] = [.in_progress]
            if actorType == .human {
                return transitions + [.done]
            }
            return transitions
        case .done, .blocked, .cancelled, .needs_info:
            return []
        }
    }
    
    // MARK: - Validation
    
    /// Validates if a transition from one state to another is allowed
    static func isValidTransition(from currentState: ItemStatus, to newState: ItemStatus, actorType: ActorType = .human) -> Bool {
        if currentState == newState {
            return false
        }
        
        if currentState == .cancelled {
            return false
        }
        
        if newState == .blocked || newState == .cancelled || newState == .needs_info {
            return true
        }
        
        if currentState == .blocked {
            return newState == .todo || newState == .in_progress
        }
        
        if currentState == .needs_info {
            return newState == .todo
        }
        
        if currentState == .done {
            return false
        }
        
        if newState == .done && actorType != .human {
            return false
        }
        
        return baseAllowedTransitions(from: currentState, actorType: actorType).contains(newState)
    }
    
    /// Asserts that a transition is valid, throwing an error if not
    static func assertValidTransition(from currentState: ItemStatus, to newState: ItemStatus, actorType: ActorType = .human) throws {
        if currentState == newState {
            throw StateMachineError.alreadyInState(currentState)
        }
        
        if currentState == .cancelled {
            throw StateMachineError.cancelledIsTerminal
        }
        
        if newState == .blocked {
            return
        }
        
        if newState == .cancelled {
            return
        }
        
        if newState == .needs_info {
            return
        }
        
        if currentState == .blocked {
            if newState == .todo || newState == .in_progress {
                return
            }
            throw StateMachineError.invalidBlockedTransition(allowed: [.todo, .in_progress], actual: newState)
        }
        
        if newState == .done && actorType != .human {
            throw StateMachineError.onlyHumanCanTransitionToDone(actorType: actorType)
        }
        
        if currentState == .needs_info {
            if newState == .todo {
                return
            }
            throw StateMachineError.invalidNeedsInfoTransition(allowed: [.todo], actual: newState)
        }
        
        if currentState == .done {
            throw StateMachineError.doneIsTerminal
        }
        
        let allowed = baseAllowedTransitions(from: currentState, actorType: actorType)
        guard allowed.contains(newState) else {
            let allowedWithGlobal = allowed + [.blocked, .cancelled, .needs_info]
            throw StateMachineError.invalidTransition(from: currentState, to: newState, allowed: allowedWithGlobal)
        }
    }
    
    // MARK: - State Properties
    
    /// Returns true if the state is a terminal state (no further transitions possible)
    static func isTerminalState(_ state: ItemStatus) -> Bool {
        return state == .done || state == .cancelled
    }
    
    /// Returns true if the state indicates active work
    static func isActiveState(_ state: ItemStatus) -> Bool {
        return state == .in_progress || state == .test
    }
    
    /// Returns the state group for UI display
    static func stateGroup(for state: ItemStatus) -> StateGroup {
        switch state {
        case .todo, .needs_info:
            return .requirement
        case .in_progress:
            return .execution
        case .test:
            return .execution
        case .done:
            return .completed
        case .blocked:
            return .blocked
        case .cancelled:
            return .cancelled
        }
    }
}

// MARK: - State Group

enum StateGroup: String, CaseIterable {
    case requirement
    case planning
    case execution
    case completed
    case blocked
    case cancelled
    
    var displayName: String {
        switch self {
        case .requirement: return "Requirement"
        case .planning: return "Planning"
        case .execution: return "Execution"
        case .completed: return "Completed"
        case .blocked: return "Blocked"
        case .cancelled: return "Cancelled"
        }
    }
}

// MARK: - Errors

enum StateMachineError: LocalizedError, Equatable {
    case alreadyInState(ItemStatus)
    case cancelledIsTerminal
    case doneIsTerminal
    case invalidBlockedTransition(allowed: [ItemStatus], actual: ItemStatus)
    case invalidNeedsInfoTransition(allowed: [ItemStatus], actual: ItemStatus)
    case onlyHumanCanTransitionToDone(actorType: ActorType)
    case invalidTransition(from: ItemStatus, to: ItemStatus, allowed: [ItemStatus])
    
    var errorDescription: String? {
        switch self {
        case .alreadyInState(let state):
            return "Item is already in state '\(state.displayName)'"
        case .cancelledIsTerminal:
            return "Cancelled items cannot be transitioned to any other state"
        case .doneIsTerminal:
            return "Items in 'done' can only transition to 'blocked', 'cancelled', or 'needs_info'"
        case .invalidBlockedTransition(let allowed, let actual):
            let allowedNames = allowed.map { $0.displayName }.joined(separator: ", ")
            return "Blocked items can only transition to \(allowedNames). Cannot transition to '\(actual.displayName)'"
        case .invalidNeedsInfoTransition(let allowed, let actual):
            let allowedNames = allowed.map { $0.displayName }.joined(separator: ", ")
            return "Items in 'needs_info' can only transition to \(allowedNames). Cannot transition to '\(actual.displayName)'"
        case .onlyHumanCanTransitionToDone(let actorType):
            return "Only human actors can transition to 'done'. Actor type '\(actorType.rawValue)' is not allowed"
        case .invalidTransition(let from, let to, let allowed):
            let allowedNames = allowed.map { $0.displayName }.joined(separator: ", ")
            return "Cannot transition from '\(from.displayName)' to '\(to.displayName)'. Allowed: \(allowedNames.isEmpty ? "none" : allowedNames)"
        }
    }
}

File: mcp-server/src/state-machine.ts
export const WORKFLOW_STATES = [
  'todo',
  'in_progress',
  'test',
  'done',
  'blocked',
  'cancelled',
  'needs_info',
] as const;

export type WorkflowState = (typeof WORKFLOW_STATES)[number];

export type ActorType = 'human' | 'ai' | 'system';

const MAIN_STATES: readonly WorkflowState[] = ['todo', 'in_progress', 'test', 'done'];
const AUXILIARY_STATES: readonly WorkflowState[] = ['blocked', 'cancelled', 'needs_info'];

/**
 * Defines allowed state transitions.
 * Key: current state (excluding auxiliary states which have special handling)
 * Value: array of allowed target states
 */
const ALLOWED_TRANSITIONS: ReadonlyMap<WorkflowState, readonly WorkflowState[]> = new Map([
  ['todo', ['in_progress']],
  ['in_progress', ['test']],
  ['test', ['done', 'in_progress']],
  ['done', []],
  ['blocked', []], // Special handling - can return to previous state
  ['cancelled', []], // Terminal state
  ['needs_info', ['todo']],
]);

function isWorkflowState(value: string): value is WorkflowState {
  return (WORKFLOW_STATES as readonly string[]).includes(value);
}

function isActorType(value: string): value is ActorType {
  return ['human', 'ai', 'system'].includes(value);
}

/**
 * Asserts that a state transition is valid.
 * 
 * @param from - Current state
 * @param to - Target state
 * @param actorType - Type of actor performing the transition
 * @param options - Additional options (previousState for blocked items)
 * @throws Error if transition is invalid
 */
export function assertValidStateTransition(
  from: string,
  to: string,
  actorType: string,
  options: { previousState?: string | null } = {},
): void {
  // Validate states
  if (!isWorkflowState(from)) {
    throw new Error(`Unknown workflow state: "${from}".`);
  }

  if (!isWorkflowState(to)) {
    throw new Error(`Unknown workflow state: "${to}".`);
  }

  if (!isActorType(actorType)) {
    throw new Error(`Unknown actor type: "${actorType}". Must be "human", "ai", or "system".`);
  }

  // Cannot transition to same state
  if (from === to) {
    throw new Error(`Item is already in state "${from}".`);
  }

  // Rule: cancelled is terminal - no transitions allowed
  if (from === 'cancelled') {
    throw new Error(`Cancelled items cannot be transitioned to any other state.`);
  }

  // Rule: Any actor can transition to blocked
  if (to === 'blocked') {
    return;
  }

  // Rule: Any actor can transition to cancelled
  if (to === 'cancelled') {
    return;
  }

  // Rule: Any actor can transition to needs_info
  if (to === 'needs_info') {
    return;
  }

  // Rule: Blocked items can return to todo or in_progress
  if (from === 'blocked') {
    if (to === 'todo' || to === 'in_progress') {
      return;
    }
    throw new Error(
      `Blocked items can only transition to "todo", "in_progress", "blocked", "cancelled", or "needs_info". Cannot transition to "${to}".`,
    );
  }

  // Rule: Only HUMAN actors can transition to done
  if (to === 'done' && actorType !== 'human') {
    throw new Error(
      `Only human actors can transition to "done". Actor type "${actorType}" is not allowed.`,
    );
  }

  // Rule: needs_info can only transition to todo
  if (from === 'needs_info') {
    if (to === 'todo') {
      return;
    }
    throw new Error(
      `Items in "needs_info" can only transition to "todo", "blocked", "cancelled", or "needs_info". Cannot transition to "${to}".`,
    );
  }

  // Rule: done is terminal (except to auxiliary states already handled above)
  if (from === 'done') {
    throw new Error(
      `Items in "done" can only transition to "blocked", "cancelled", or "needs_info". Cannot transition to "${to}".`,
    );
  }

  // Check allowed transitions for main states
  const allowedTargets = ALLOWED_TRANSITIONS.get(from) ?? [];
  if (allowedTargets.includes(to)) {
    return;
  }

  // Build allowed description for error message
  const allowedDescription = [...allowedTargets, ...AUXILIARY_STATES];
  throw new Error(
    `Cannot transition item from "${from}" to "${to}". Allowed states: ${allowedDescription.join(', ')}.`,
  );
}

/**
 * Gets all allowed transitions for a given state.
 * 
 * @param state - Current state
 * @param actorType - Actor type (affects whether 'done' is included)
 * @returns Array of allowed target states
 */
export function getAllowedTransitions(
  state: WorkflowState,
  actorType: ActorType = 'human',
): WorkflowState[] {
  // Cancelled is terminal
  if (state === 'cancelled') {
    return [];
  }

  // Done can only go to auxiliary states
  if (state === 'done') {
    return ['blocked', 'cancelled', 'needs_info'];
  }

  // Blocked can go to todo, in_progress, or auxiliary states
  if (state === 'blocked') {
    return ['todo', 'in_progress', 'cancelled', 'needs_info'];
  }

  // needs_info can go to todo or auxiliary states
  if (state === 'needs_info') {
    return ['todo', 'blocked', 'cancelled'];
  }

  // Main states
  const baseTransitions = ALLOWED_TRANSITIONS.get(state) ?? [];
  
  // Filter out 'done' if actor is not human
  const filteredTransitions = baseTransitions.filter(t => {
    if (t === 'done') {
      return actorType === 'human';
    }
    return true;
  });

  return [...filteredTransitions, ...AUXILIARY_STATES];
}
